Introduce a new env variable to allow spaces in home path (#1446)
pypa/pipx: docs/installation.md

If you wish to move to a different location, just replace the `NEW_LOCATION` value.

### MacOS

Current default location: `~/.local`

```bash
NEW_LOCATION=~/.local
cache_dir=$(pipx environment --value PIPX_VENV_CACHEDIR)
logs_dir=$(pipx environment --value PIPX_LOG_DIR)
trash_dir=$(pipx environment --value PIPX_TRASH_DIR)
home_dir=$(pipx environment --value PIPX_HOME)
rm -rf "$cache_dir" "$logs_dir" "$trash_dir"
mkdir -p $NEW_LOCATION && mv "$home_dir" $NEW_LOCATION
pipx reinstall-all
```

### Linux

Current default location: `~/.local/share`

```bash
cache_dir=$(pipx environment --value PIPX_VENV_CACHEDIR)
logs_dir=$(pipx environment --value PIPX_LOG_DIR)
trash_dir=$(pipx environment --value PIPX_TRASH_DIR)
home_dir=$(pipx environment --value PIPX_HOME)
# If you wish another location, replace the if block below
# and set `NEW_LOCATION` explicitly
if [ -n "$XDG_DATA_HOME" ]; then
    NEW_LOCATION="$XDG_DATA_HOME"
else
    NEW_LOCATION="$HOME/.local/share"
fi
rm -rf $cache_dir $logs_dir $trash_dir
mkdir -p $NEW_LOCATION && mv $home_dir $NEW_LOCATION
pipx reinstall-all
```

### Windows

Requires bash (for example `git bash`)

> Current default location: `~`

Reviewer: ```suggestion
Current default location: `~/pipx`
```

According to the implementation here [1], the default location should be "~/pipx".

[1] https://github.com/pypa/pipx/blob/dc9f9a849a918c68bc2cc99c72651022371fcf9f/src/pipx/paths.py#L16

> ```bash
> NEW_LOCATION=~
> cache_dir=$(pipx environment --value PIPX_VENV_CACHEDIR)
> logs_dir=$(pipx environment --value PIPX_LOG_DIR)
> trash_dir=$(pipx environment --value PIPX_TRASH_DIR)
> home_dir=$(pipx environment --value PIPX_HOME)
> rm -rf $cache_dir $logs_dir $trash_dir
> mv $home_dir $NEW_LOCATION
> pipx reinstall-all
> ```

Reviewer: ````suggestion
```powershell
$NEW_LOCATION = Join-Path $HOME '.local/pipx'
$cache_dir = pipx environment --value PIPX_VENV_CACHEDIR
$logs_dir = pipx environment --value PIPX_LOG_DIR
$trash_dir = pipx environment --value PIPX_TRASH_DIR
$home_dir = pipx environment --value PIPX_HOME

Remove-Item -Recurse -Force -ErrorAction SilentlyContinue $cache_dir, $logs_dir, $trash_dir

Move-Item -Path $home_dir -Destination $NEW_LOCATION
pipx reinstall-all
```
````

BTW, the process works fine on my PC.

Author: In powershell, the `Move-Item` command moves the files in the specified paths into the `-Destination` path?

Reviewer: Good catch! I asked ChatGPT and tested on my PC, it seems to depend on the existence of destination directory.

If the destination directory exists, the behavior is just like bash. When the destionation does not exists, it works more like a rename. That is why it works on my side previously.

So maybe we should remove the destionation directory before `Move-Item` to ensure the rename behavior. WDYT?

BTW, I updated the suggested changes accordingly.

Author: Gotcha, yeah nice this looks good to me. Thanks for helping me out here, I'm useless when it comes to windows ;)

```
pipx completions
```

## Moving your pipx installation

The below code snippets show how to move your pipx installation to a new directory.
As an example, they move from a non-default location to the current default locations.
If you wish to move to a different location, just replace the `NEW_LOCATION` value.

### MacOS

Current default location: `~/.local`

```bash
NEW_LOCATION=~/.local
cache_dir=$(pipx environment --value PIPX_VENV_CACHEDIR)
logs_dir=$(pipx environment --value PIPX_LOG_DIR)
trash_dir=$(pipx environment --value PIPX_TRASH_DIR)
home_dir=$(pipx environment --value PIPX_HOME)
rm -rf "$cache_dir" "$logs_dir" "$trash_dir"
mkdir -p $NEW_LOCATION && mv "$home_dir" $NEW_LOCATION
pipx reinstall-all
```

### Linux

Current default location: `~/.local/share`

```bash
cache_dir=$(pipx environment --value PIPX_VENV_CACHEDIR)
logs_dir=$(pipx environment --value PIPX_LOG_DIR)
trash_dir=$(pipx environment --value PIPX_TRASH_DIR)
home_dir=$(pipx environment --value PIPX_HOME)

> # If you wish another location, replace the if block below
> # and set `NEW_LOCATION` explicitly
> if [ -n "$XDG_DATA_HOME" ]; then
>     NEW_LOCATION="$XDG_DATA_HOME"
> else
>     NEW_LOCATION="$HOME/.local/share"
> fi

Reviewer 2: We might be able to simplify this:

```suggestion
# If you wish another location, replace the expression below
# and set `NEW_LOCATION` explicitly
NEW_LOCATION="${XDG_DATA_HOME:-$HOME/.local/share}"
```

rm -rf "$cache_dir" "$logs_dir" "$trash_dir"
mkdir -p $NEW_LOCATION && mv "$home_dir" $NEW_LOCATION
pipx reinstall-all
```

### Windows

> Requires bash (for example `git bash`)

Reviewer 2: ```suggestion
Requires bash (for example `git bash`).
```

Current default location: `~/pipx`

```powershell
$NEW_LOCATION = Join-Path "$HOME" 'pipx'
$cache_dir = pipx environment --value PIPX_VENV_CACHEDIR
$logs_dir = pipx environment --value PIPX_LOG_DIR
$trash_dir = pipx environment --value PIPX_TRASH_DIR
$home_dir = pipx environment --value PIPX_HOME

Remove-Item -Recurse -Force -ErrorAction SilentlyContinue "$cache_dir", "$logs_dir", "$trash_dir"

# Remove the destination directory to ensure rename behavior of `Move-Item`
Remove-Item -Recurse -Force -ErrorAction SilentlyContinue "$NEW_LOCATION"

Move-Item -Path $home_dir -Destination "$NEW_LOCATION"
pipx reinstall-all
```

> If you would prefer doing it in bash via git-bash/WSL, feel free to use

Reviewer 2: Don't the instructions say that git-bash is required, above?

Author: This was a leftover from what I copied from the discussion. Removed it.
